### PR TITLE
Incrementally push to the scrub table to reduce memory pressure

### PR DESCRIPTION
**Pull request by CVdV-au**
**Goals (and why)**:

Reduce memory pressure when large number of cells are deleted.

**Implementation Description (bullets)**:
Incrementally persist the cells to delete, rather than generating all the data and then incrementally persisting.
See PDS-190294 where a 20K object realm delete causes a GC storm with a client heap of 64G

**Testing (What was existing testing like?  What have you done to improve it?)**:
N/A

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServiceScrubberStore.java [hunk 1]
```diff
@@ -114,14 +113,14 @@ public void queueCellsForScrubbing(
                 byte[] tableBytes = EncodingUtils.encodeVarString(tableRef.getQualifiedName());
                 byte[] col = EncodingUtils.add(tableBytes, cell.getColumnName());
                 values.put(Cell.create(cell.getRowName(), col), EMPTY_CONTENTS);
```
[the remark this answers is not in the document]
Comment (schlosna) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServiceScrubberStore.java:115]: on line 108, if we're concerned about allocations, might want to presize the `values` buffer Map:
```
Map<Cell, byte[]> values = Maps.newHashMapWithExpectedSize(Math.min(batchSize, cellToTableRefs.size()));
```

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServiceScrubberStore.java [hunk 2]
```diff
@@ -114,14 +113,14 @@ public void queueCellsForScrubbing(
                 byte[] tableBytes = EncodingUtils.encodeVarString(tableRef.getQualifiedName());
                 byte[] col = EncodingUtils.add(tableBytes, cell.getColumnName());
```
[the remark this answers is not in the document]
Comment (schlosna) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServiceScrubberStore.java:115]: There is a `com.palantir.atlasdb.encoding.PtBytes#toCachedBytes` but that is typically used for small compile time strings like column name `PtBytes.toCachedBytes("i")` so might not be worth it.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServiceScrubberStore.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (CVdV-au) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServiceScrubberStore.java:121]: Yes, that is the intent of the change. There are only downsides to computing all the rows we are going to push to storage, and then pushing - we should "push as we go".

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServiceScrubberStore.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (CVdV-au) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServiceScrubberStore.java]: We could probably have a single cache but keeping them separate seems reasonable given there will never be an overlap in keys (we'd have to convert the cache to Object->byte[] or convert TableReference->String). What you've done seems reasonable.

**Review by schlosna — APPROVED** [review]

File: atlasdb-impl-shared/build.gradle [hunk 3]
```diff
@@ -17,6 +17,7 @@ dependencies {
   compile project(":timestamp-api")
   compile project(":timestamp-client")
 
+  compile 'com.github.ben-manes.caffeine:caffeine'
```
Comment (jeremyk-91) [inline atlasdb-impl-shared/build.gradle:20]: nit: can we do `implementation`? `compile` is the legacy thing

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServiceScrubberStore.java [hunk 4]
```diff
@@ -71,6 +72,10 @@ protected String getInitializingClassName() {
     private static final byte[] EMPTY_CONTENTS = new byte[] {1};
     private final InitializingWrapper wrapper = new InitializingWrapper();
     private final KeyValueService keyValueService;
+    private final LoadingCache<TableReference, byte[]> tableToEncodedBytesCache = Caffeine.newBuilder()
```
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServiceScrubberStore.java:75]: nice!

**Review by jeremyk-91 — APPROVED** [review]
👍  looks good

File: atlasdb-impl-shared/build.gradle (hunk 3 above)
Comment (schlosna) [reply atlasdb-impl-shared/build.gradle:20]: Will push a follow up PR to cleanup a bunch of the gradle dependencies

**Pull request merged**